src/gui/torrentcontentmodel.cpp

```diff
@@ -497,33 +530,25 @@ void TorrentContentModel::setupModelData(const BitTorrent::AbstractFileStorage &
     qDebug("Torrent contains %d files", filesCount);
     m_filesIndex.reserve(filesCount);
 
-    TorrentContentModelFolder *currentParent;
+    FolderAdaptor::FolderAdaptorPtr rootFolderAdaptor = std::make_shared<FolderAdaptor>(m_rootItem);
+    FolderAdaptor::FolderAdaptorPtr currentParent = nullptr;
     // Iterate over files
     for (int i = 0; i < filesCount; ++i)
     {
-        currentParent = m_rootItem;
+        currentParent = rootFolderAdaptor;
         const QString path = Utils::Fs::toUniformPath(info.filePath(i));
 
         // Iterate of parts of the path to create necessary folders
         QList<QStringView> pathFolders = QStringView(path).split(u'/', Qt::SkipEmptyParts);
         pathFolders.removeLast();
 
-        for (const QStringView pathPart : asConst(pathFolders))
-        {
-            const QString folderPath = pathPart.toString();
-            TorrentContentModelFolder *newParent = currentParent->childFolderWithName(folderPath);
-            if (!newParent)
-            {
-                newParent = new TorrentContentModelFolder(folderPath, currentParent);
-                currentParent->appendChild(newParent);
-            }
-            currentParent = newParent;
-        }
+        for (const QStringView folderName : asConst(pathFolders))
+            currentParent = currentParent->findChildFolderOrInsert(folderName.toString());
+
         // Actually create the file
-        TorrentContentModelFile *fileItem = new TorrentContentModelFile(
```

Comment:
I don't understand why move this part into adaptor class. Isn't the adaptor only for speeding up look ups?

Reply:
`TorrentContentModelFolder::appendChild` can be dangerous if used with FolderAdaptor

Reply:
The adaptor class seems so strange to me... https://github.com/qbittorrent/qBittorrent/pull/16183#discussion_r789697508